README.md
GreenClock
==========

[![Travs-CI status](https://travis-ci.org/pcdinh/greenclock.png)](https://travis-ci.org/pcdinh/greenclock) 
[![Coverage Status](https://coveralls.io/repos/pcdinh/greenclock/badge.png)](https://coveralls.io/r/pcdinh/greenclock)

GreenClock is a time-based task scheduler using gevent

With GreenClock, you can: Schedule a task to run every X seconds, daily, weekly, monthly, 
or at certain times (such as application startup).

GreenClock launches a [green thread](http://en.wikipedia.org/wiki/Green_threads) per task.
Therefore every task will be executed in a concurrent manner, without blocking each other.

### Status

This module is currently under development.

### Features

    - A simple to use API for scheduling jobs.
    - Lightweight (depending on gevent only)
    - Works with Python 2.7+
    - Support the following scheduling scenarios: 
        + run every X seconds
        + run every hour at specified minute and second
        + run at specified time (hour:minute:second) every day
        + more to come

### Installation

This library depends on gevent 1.5

```bash
    $ pip install cython -e git://github.com/surfly/gevent.git@1.0rc2#egg=gevent
```

To install GreenClock from [pip](https://pypi.python.org/pypi/pip):

```bash
    $ pip install greenclock
```


You can also install it into your Python application directory

```bash
    $ pip install --install-option="--prefix=/path/to/python/app" greenclock
```


To install GreenClock from source:
```bash
    $ git clone git@github.com:pcdinh/greenclock.git
    $ python setup.py install
```

### Usage

```
from greenclock.utils import Scheduler
from datetime import datetime
import time

def func_1():
    print('Calling func_1() at ' + str(datetime.now()))
    time.sleep(2)
    print('Ended call to func_1() at ' + str(datetime.now()))

def func_2():
    print('Calling func_2() at ' + str(datetime.now()))
    time.sleep(2)
    print('Ended call to func_2() at ' + str(datetime.now()))

if __name__ == "__main__":
    scheduler = Scheduler(logger_name='task_scheduler')
    scheduler.schedule('task_1', greenclock.every_second(4), func_1)
    scheduler.schedule('task_2', greenclock.every_second(1), func_2)
    # Run hourly task at 41:00 every day
    scheduler.schedule('task_3', greenclock.every_hour(minute=41, second=0), func_3)
    # Run daily task at 12:35:00
    scheduler.schedule('task_2', greenclock.every_hour(hour=12, minute=35, second=0), func_2)    
    # To start the scheduled tasks immediately, specify 'once' for `start_at`
    # Other values: 
    # * `next_minute`: Wait until the first seconds of the next minute to run
    # * `next_hour`: Wait until the first seconds of the next hour to run
    # * `tomorrow`: Wait until the first seconds of tomorrow to run
    scheduler.run_forever(start_at='once')

```

Basically to schedule a periodic task or job, you need to specify the following parameters:

    + Task name: `task_1`
    + A timer that let the scheduler know how to run a periodic task

```        
        # run the task for every 4 seconds
        from greenclock.utils import every_second, every_hour
        every_second(4) 
        # run the task every day at 01:10:00
        every_hour(hour=1, minute=10, second=0)
```        
    + A function or callable object
    + Optional parameters to the above function or callable object
```    
        scheduler.schedule('task_1', greenclock.every_second(1), func_1, param1, param2, named_param=2)
```    
`Scheduler` object can run a separate process which never exits if you want it to
```
    scheduler.run_forever(start_at='once')
```

[![Bitdeli Badge](https://d2weczhvl823v0.cloudfront.net/pcdinh/greenclock/trend.png)](https://bitdeli.com/free "Bitdeli Badge")
[![githalytics.com alpha](https://cruel-carlota.pagodabox.com/a7b875db36121c410c906c620f242458 "githalytics.com")](http://githalytics.com/pcdinh/greenclock)


CHANGELOG.md

# History
---------

0.3.0 (2020-03-03)
++++++++++++++++++

- Initial support for Python 3.8 and gevent 1.5

0.1.0 (2013-08-09)
++++++++++++++++++

- Initial release

tests/__init__.py
.. :changelog:

History
-------

0.1.0 (2013-08-09)
++++++++++++++++++

- Initial release

tests/test_scheduling.py
# coding: utf-8

import datetime
import unittest
from gevent.event import AsyncResult
from gevent import sleep
from greenclock.utils import every_second, every_hour, Scheduler


INTERVAL = 2


class MockDate1(datetime.datetime):
    @classmethod
    def now(cls):
        return cls(2013, 8, 11, 0, 0, 0)


class MockDatetime(object):

    def __init__(self, new_datetime):
        self.old_datetime = datetime.datetime
        self.new_datetime_cls = new_datetime

    def __enter__(self):
        datetime.datetime = self.new_datetime_cls
        return self

    def __exit__(self, type, value, traceback):
        datetime.datetime = self.old_datetime
        return self


def func_1():
    return 100


class SchedulerTests(unittest.TestCase):

    def setUp(self):
        self.scheduler = Scheduler()

    def tearDown(self):
        self.scheduler.stop()

    def test_builtin_timer_second(self):
        assert every_second(2).__next__().total_seconds() == 2
        assert every_second(3600).__next__().total_seconds() == 3600

    def test_builtin_timer_hour1(self):
        # Unable to mock the datetime.datetime.now using @mock.patch('datetime.datetime.now')
        # See http://stackoverflow.com/questions/4481954/python-trying-to-mock-datetime-date-today-but-not-working
        # datetime.datetime.now.return_value = datetime.datetime(2013, 8, 11, 0, 0, 0)
        with MockDatetime(MockDate1):
            timer = every_hour(minute=10, second=0)
            # First time
            delay1 = timer.next().total_seconds()
            assert delay1 == 10 * 60, 'Actual %s' % str(delay1)
            assert timer.next().total_seconds() == 60 * 60

    def test_builtin_timer_hour2(self):
        with MockDatetime(MockDate1):
            timer = every_hour(hour=1, minute=10, second=0)
            # First time
            delay1 = timer.next().total_seconds()
            assert delay1 == 60 + 10 * 60, 'Actual %s' % str(delay1)
            assert timer.next().total_seconds() == 24 * 60 * 60

    def test_task_assignment(self):
        # See https://github.com/surfly/gevent/blob/master/greentest/test__event.py
        self.scheduler.schedule('task_1', every_second(INTERVAL), func_1)
        assert self.scheduler.tasks[0].name == 'task_1'
        greenlet1, greenlet2 = self.scheduler.run(self.scheduler.tasks[0])
        event1 = AsyncResult()
        greenlet1.link(event1)
        self.assertEqual(event1.get(), 100)
        assert greenlet2.ready() is False
        assert greenlet2.successful() is False
        sleep(INTERVAL)
        assert greenlet2.ready() is True
        assert greenlet2.successful() is True
        greenlet2.kill()
        greenlet1.kill()
        assert greenlet2.value[0].get() == 100, 'Actual %s' % str(greenlet2.value[0].get())


if __name__ == '__main__':
    unittest.main()


greenclock/utils.py
import types
import logging
import time
from datetime import timedelta, datetime
import gevent
from gevent.pool import Pool
from gevent import monkey

monkey.patch_all()


def every_second(seconds):
    """
    Iterator-based timer

    @example
    >> every(seconds=10)
    :return an iterator of timedelta object
    """
    delta = timedelta(seconds=seconds)
    # Never return StopIteration
    while 1:
        yield delta


class every_hour(object):
    """
    A class-based iterator that help install a timer for hourly scheduled task
    - Every hour in a day
    - Fixed hour in a day

    The name is chosen in all lower case to make it looks like a function because it will
    be used as if it was a generator.
    """

    def __init__(self, hour=None, minute=0, second=0):
        """
        :param hour:
        :param minute:
        :param second:
        """
        self.started = False
        self.hour = hour
        self.minute = minute
        self.second = second

    def __iter__(self):
        return self

    def next(self):
        """
        Never return StopIteration
        """
        if self.started is False:

            self.started = True
            time_now = datetime.now()

            if self.hour == 0 or self.hour:
                # Fixed hour in a day
                # Next run will be the next day
                scheduled = time_now.replace(hour=self.hour, minute=self.minute, second=self.second, microsecond=0)
                if scheduled == time_now:
                    return timedelta(seconds=0)
                if scheduled < time_now:
                    # Scheduled time is passed
                    return scheduled + timedelta(days=1) - time_now
            else:
                # Every hour in a day
                # Next run will be the next hour
                scheduled = time_now.replace(minute=self.minute, second=self.second, microsecond=0)
                if scheduled == time_now:
                    return timedelta(seconds=0)
                if scheduled < time_now:
                    # Scheduled time is passed
                    return scheduled + timedelta(hour=1) - time_now
            return scheduled - time_now
        else:
            if self.hour:
                return timedelta(days=1)  # next day
            return timedelta(hours=1)  # next hour


def wait_until(time_label):
    """
    Calculates the number of seconds that the process needs to sleep
    """
    if time_label == 'next_minute':
        gevent.sleep(60 - int(time.time()) % 60)
    elif time_label == 'next_hour':
        gevent.sleep(3600 - int(time.time()) % 3600)
    elif time_label == 'tomorrow':
        gevent.sleep(86400 - int(time.time()) % 86400)


class Task(object):
    """A scheduled task"""

    def __init__(self, name, action, timer, *args, **kwargs):
        self.name = name
        self.action = action
        self.timer = timer
        self.args = args
        self.kwargs = kwargs


class Scheduler(object):
    """
    Time-based scheduler
    """

    def __init__(self, logger_name='greenlock.task'):
        self.logger_name = logger_name
        self.tasks = []
        self.active = {}  # action task name registry
        self.waiting = {}  # action task name registry
        self.running = True

    def schedule(self, name, timer, func, *args, **kwargs):
        """
        ts = Scheduler('my_task')
        ts.schedule(every(seconds=10), handle_message, "Every 10 seconds")
        ts.schedule(every(seconds=30), fetch_url, url="http://yahoo.com", section="stock_ticker")
        ts.run_forever()
        """
        self.tasks.append(Task(name, func, timer, *args, **kwargs))
        self.active[name] = []  # list of greenlets
        self.waiting[name] = []

    def unschedule(self, task_name):
        """
        Removes a task from scheduled jobs but it will not kill running tasks
        """
        for greenlet in self.waiting[task_name]:
            try:
                gevent.kill(greenlet)
            except BaseException:
                pass

    def stop_task(self, task_name):
        """
        Stops a running or dead task
        """
        for greenlet in self.active[task_name]:
            try:
                # Do not need to check if greenlet is dead, gevent does it already
                gevent.kill(greenlet)
                self.active[task_name] = []
            except BaseException:
                pass

    def _remove_dead_greenlet(self, task_name):
        """
        Removes dead greenlet or done task from active list
        """
        for greenlet in self.active[task_name]:
            try:
                # Allows active greenlet continue to run
                if greenlet.dead:
                    self.active[task_name].remove(greenlet)
            except BaseException:
                pass

    def run(self, task):
        """
        Runs a task and re-schedule it
        """
        self._remove_dead_greenlet(task.name)
        if isinstance(task.timer, types.GeneratorType):
            # Starts the task immediately
            green_thread = gevent.spawn(task.action, *task.args, **task.kwargs)
            self.active[task.name].append(green_thread)
            try:
                # total_seconds is available in Python 2.7
                green_thread_later = gevent.spawn_later(task.timer.__next__().total_seconds(), self.run, task)
                self.waiting[task.name].append(green_thread_later)
                return green_thread, green_thread_later
            except StopIteration:
                pass
            return green_thread, None
        # Class based timer
        try:
            if task.timer.started is False:
                delay = task.timer.next().total_seconds()
                gevent.sleep(delay)
                green_thread = gevent.spawn(task.action, *task.args, **task.kwargs)
                self.active[task.name].append(green_thread)
            else:
                green_thread = gevent.spawn(task.action, *task.args, **task.kwargs)
                self.active[task.name].append(green_thread)
            green_thread_later = gevent.spawn_later(task.timer.__next__().total_seconds(), self.run, task)
            self.waiting[task.name].append(green_thread_later)
            return green_thread, green_thread_later
        except StopIteration:
            pass
        return green_thread, None

    def run_tasks(self):
        """
        Runs all assigned task in separate green threads. If the task should not be run, schedule it
        """
        pool = Pool(len(self.tasks))
        for task in self.tasks:
            # Launch a green thread to schedule the task
            # A task will be managed by 2 green thread: execution thread and scheduling thread
            pool.spawn(self.run, task)
        return pool

    def run_forever(self, start_at='once'):
        """
        Starts the scheduling engine

        :param start_at: 'once' -> start immediately
                         'next_minute' -> start at the first second of the next minutes
                         'next_hour' -> start 00:00 (min) next hour
                         'tomorrow' -> start at 0h tomorrow
        """
        if start_at not in ('once', 'next_minute', 'next_hour', 'tomorrow'):
            raise ValueError(
                "start_at parameter must be one of these values: 'once', 'next_minute', 'next_hour', 'tomorrow'")
        if start_at != 'once':
            wait_until(start_at)
        try:
            task_pool = self.run_tasks()
            while self.running:
                gevent.sleep(seconds=1)
            task_pool.join(timeout=30)
            task_pool.kill()
        except KeyboardInterrupt:
            # https://github.com/surfly/gevent/issues/85
            task_pool.closed = True
            task_pool.kill()
            logging.getLogger(self.logger_name).info('Time scheduler quits')

    def stop(self):
        """
        Stops the scheduling engine
        """
        self.running = False


greenclock/__init__.py
# coding: utf-8
# license: MIT
"""A library to provide an interface to schedule and manage periodic tasks using `gevent`.
To use the library, you need to import the module first    
    import greenclock
    
Tasks are managed by `Scheduler` objects
    
    scheduler = greenclock.Scheduler(logger_name='task_scheduler')
    
You can register scheduled tasks with `Scheduler` object

    scheduler.schedule('task_1', greenclock.every_second(4), func_1)
    scheduler.schedule('task_2', greenclock.every_second(1), func_2)
    
Basically to schedule a periodic task or job, you need to specify the following parameters:

    + Task name: `task_1`
    + A timer that let the scheduler know how to run a periodic task
        
        # run the task for every 4 seconds
        greenclock.every_second(4) 
        # run the task every day at 01:10:00
        greenclock.every_hour(hour=1, minute=10, second=0)
        
    + A function or callable object
    + Optional parameters to the above function or callable object
    
        scheduler.schedule('task_2', greenclock.every_second(1), func_2, param1, param2, named_param=2)
    
`Scheduler` object can run a separate process which never exits if you want it to

    scheduler.run_forever(start_at='once')
"""

from . import version
__version__ = version.__version__


example.py

from greenclock.utils import Scheduler, every_second, every_hour
from datetime import datetime
import time


def func_1():
    print('Calling func_1() at ' + str(datetime.now()))
    time.sleep(2)
    print('Ended call to func_1() at ' + str(datetime.now()))


def func_2():
    print('Calling func_2() at ' + str(datetime.now()))
    time.sleep(2)
    print('Ended call to func_2() at ' + str(datetime.now()))


def func_3():
    print('Calling func_3() at ' + str(datetime.now()))
    time.sleep(2)
    print('Ended call to func_3() at ' + str(datetime.now()))


if __name__ == "__main__":
    scheduler = Scheduler(logger_name='task_scheduler')
    scheduler.schedule('task_1', every_second(4), func_1)
    scheduler.schedule('task_2', every_second(1), func_2)
    # Run at 41:00 every hour every day
    scheduler.schedule('task_3', every_hour(minute=41, second=0), func_3)
    # Run task at 12:35:00 every day
    scheduler.schedule('task_2', every_hour(hour=12, minute=35, second=0), func_2)
    # Starts the scheduling engine without delay
    scheduler.run_forever(start_at='once')
